receiver/carbonreceiver/protocol/regex_parser.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package protocol
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+const (
+	metricNameCapturePrefix = "name_part"
+)
+
+// RegexParserConfig has the configuration for a parser that can breakdown a
+// Carbon "metric path" and transform it in corresponding metric labels according
+// to a series of regular expressions rules (see below for details).
+//
+// This is typically used to extract labels from a "naming hierarchy", see
+// https://graphite.readthedocs.io/en/latest/feeding-carbon.html#step-1-plan-a-naming-hierarchy
+//
+// Examples:
+//
+// 1. Rule:
+//        - regexp: "(?P<svc>[^.]+)\.(?P<host>[^.]+)\.cpu\.seconds"
+//          name_prefix: cpu_seconds
+//          labels:
+//            k: v
+//    Metric path: "service_name.host00.cpu.seconds"
+//    Resulting metric:
+//        name: cpu_seconds
+//        label keys: {"svc", "host", "k"}
+//        label values: {"service_name", "host00", "k"}
+//
+// 2. Rule:
+//        - regexp: "^(?P<svc>[^.]+)\.(?P<host>[^.]+)\.(?P<name_part0>[^.]+).(?P<name_part1>[^.]+)$"
```

Comment:
Would it be worth adding a `metric_` or `key_` prefix to params other than the name? Might make it a bit more obvious and explicit. So this will look like `"^(?P<metric_svc>[^.]+)\.(?P<metric_host>[^.]+)\.(?P<name_0>[^.]+).(?P<name_1>[^.]+)$"`. Any matches found with `name_` prefix continue to be concatenated into the name automatically and anything with a `metric_` prefix gets extracted into key/value pairs. All other will be  params are ignored.

Reply:
It will have to error out if doesn't have a prefix but it seems "self-documenting", the only downside is that the regexp gets a bit bigger, but it seems worth it.